### PR TITLE
Issue #2166: Add 'requiredTranslations' option to Translation check

### PR DESCRIPTION
**Pull request by MEZk**

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 1]
```diff
@@ -68,9 +72,17 @@
  * <p>
  * app-dev.properties //separator is '-'
  * </p>
+ *
+ * <p><b>checkedLanguageCodes</b> which allows to specify language codes of
+ * translation files which must exist in project. For example, if you need
+ * to check that Japanese and French localization files are present in your project,
+ * you should set checkedLanguageCodes to "ja, fr".
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: If you'll set it like here, will the check also accept other translations? Please clarify in docs.
Also, what about ROOT translation (no code in file name)?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 2]
```diff
@@ -91,6 +110,11 @@
     private String basenameSeparator;
 
     /**
+     * Language codes of translation files (de, pt, ja, etc).
+     */
+    private SortedSet<String> checkedLanguageCodes = Sets.newTreeSet();
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: Can we create here just empty set?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 3]
```diff
@@ -114,10 +147,75 @@ public void finishProcessing() {
...
         checkPropertyFileSets(propFilesMap);
     }
 
     /**
+     * Checks existence of the translation files (arranged in a map)
+     * for each resource bundle in project.
+     * @param translationFilesMap the translation files bundles organized as Map.
+     */
+    private void checkExistenceOfTranslationFiles(Map<String, Set<File>> translationFilesMap) {
+        final Set<Entry<String, Set<File>>> resourceBundles = translationFilesMap.entrySet();
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: Maybe we can inline this variable? Type for it is quite complex :)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 4]
```diff
@@ -114,10 +147,75 @@ public void finishProcessing() {
...
+        for (Entry<String, Set<File>> bundle : resourceBundles) {
+            final SortedSet<File> filesInResourceBundle = new TreeSet<>(bundle.getValue());
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: Can you explain why we need to create one more set here? Can't it be proper `TreeSet` from the very beginning?

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_ja.properties [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+# for DefaultLogger.java
+#
+DefaultLogger.addException=エラー監査 {0}
+DefaultLogger.auditStarted=監査を開始しています...
+DefaultLogger.auditFinished=監査は行われ。
+
+PackageObjectFactory.unableToInstantiateExceptionMessage=''{0}'' クラスをインスタンス化することができません、\
+  それは としてインスタンス化することもできません {1}. \
+  そのクラス名が正規名として指定されている再確認または短い名前の使用 \
+  http://checkstyle.sourceforge.net/config.html#Packages を設定する方法をお読みください。\
```
Comment (mkordas) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_ja.properties]: We need to escape `#` here, because otherwise it starts comment. Please also recheck other property files

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 6]
```diff
@@ -68,9 +72,17 @@
  * <p>
  * app-dev.properties //separator is '-'
  * </p>
+ *
+ * <p><b>checkedLanguageCodes</b> which allows to specify language codes of
+ * translation files which must exist in project. For example, if you need
+ * to check that Japanese and French localization files are present in your project,
+ * you should set checkedLanguageCodes to "ja, fr".
+ * Default value is <b>{}</b> which means that the existence of
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: What do you mean by `{}` here? It is neither empty string nor empty set.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 7]
```diff
@@ -567,4 +567,46 @@ public void testListFilesDirectoryWithNull() throws Exception {
         List<File> result = (List<File>) method.invoke(null, fileMock);
         assertEquals(0, result.size());
     }
+
+    @Test
+    public void testTranslationCheckMissinTranslationFiles() throws Exception {
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Why do we have this test in `MainTest`? It should belong to `TranslationCheckTest`

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 8]
```diff
@@ -567,4 +567,46 @@ public void testListFilesDirectoryWithNull() throws Exception {
...
+        exit.expectSystemExitWithStatus(6);
+        final String platforIndependentPath = "src" + File.separator + "test"
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: You can use `String.format` here to not repeat `File.separator` so many times

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 7 above)
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: `Missin` -> `Missing`

**Comment by mkordas** [comment]
@MEZk 
I'm done with my code review, just two more things:

1. `checkedLanguageCodes` will become our API, so we should choose this name carefully. Are you sure about this name? Maybe you can propose some other options and we'll do some voting? My proposition would be `requiredTranslations`.
2. We should correct grammar in commit message. I propose `Issue #2166: Add 'checkedLanguageCodes' option to Translation check` instead of `Issue #2166: Add check of existence of translation files for TranslationCheck`.

**Comment by MEZk** [comment]
@mkordas 
Thank you for so detailed review, I'll provide changes ASAP.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 3 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
If I make it like this:

````java
private void checkExistenceOfTranslationFiles(Map<String, SortedSet<File>> translationFilesMap) {
    for (String baseName : translationFilesMap.keySet()) {
        final SortedSet<File> filesInResourceBundle = translationFilesMap.get(baseName);
    checkExistenceOfTranslationFilesInResourceBundle(filesInResourceBundle);
    }
}
````
then findbugs will rise violation.

```
[INFO] com.puppycrawl.tools.checkstyle.checks.TranslationCheck.checkExistenceOfTranslationFiles(Map) makes inefficient use of keySet iterator instead of entrySet iterator [com.puppycrawl.tools.checkstyle.checks.TranslationCheck] At TranslationCheck.java:[line 164]
```
I don't sure how to make it simpler.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 4 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
Is it a good idea to change
```Map<String, Set<File>> translationFilesMap``` to ```Map<String, SortedSet<File>> translationFilesMap``` and use it in the whole TranslationCheck class?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 3 above)
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: Use entrySet instead keyset to iterate over map

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 4 above)
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: Yes, I think it's good idea

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 2 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
Yes. why not? Sets.newTreeSet() returns a new, empty set. By default we cannot imagine what kind translations are needed.
Collections.EMPTY_SORTED_SET is not supported in Java 1.7 (marked as since 1.8).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 1 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
Corrected javadoc/xdoc. Added check for default translation.
The name of default translation file begins with the base name of the ResourceBundle and ends with the .properties suffix.
FYI: https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 3 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 4 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
Done.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_ja.properties (hunk 5 above)
Comment (MEZk) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_ja.properties]: @mkordas 
Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java (hunk 6 above)
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: @mkordas 
I mean empty string set. Corrected javadoc/xdoc.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 7 above)
Comment (MEZk) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @mkordas 
Movet to TranslationCheckTest.
Corrected typo.

**Comment by MEZk** [comment]
@mkordas 
>My proposition would be requiredTranslations.

I support your idea.

2) Corrected.

**Comment by mkordas** [comment]
@MEZk 
I've provided detailed review inline in the commits. We're much closer to the finish now. Please should if you disagree with something.

**Comment by MEZk** [comment]
@mkordas 
>Link to ISO 639-1 language codes should be part of some sentence

Corrected.

>This Javadoc is the same as for field above - they should be different as they describe different fields I believe.

Corrected.

>We don't need to allocate any memory for TreeSet here, so I propose to use ImmutableSortedSet.of(); instead.

Replaced with ImmutableSortedSet.of()

>This is duplicated code: ... Please extract it to a method

Done.

>Duplicated operation: .substring(baseName.lastIndexOf(File.separator) + 1).
Please extract it to a private static method.

Done.

>I think is missing should be used instead of is missed.

Corrected in all translation files.

>Whenever we will change anything in Main class this test will start failing.
Can't we create just unit test for TranslationCheck? Please explain why it is not possible to avoid using Main for this test.

My bad. We can create just unit test for TranslationCheck.

>Do we really need to configure charset, 'severityandfileExtensions` for this test? It's not unit test anymore...

I've deleted the file, since it is not necessary anymore.

>1) It's link to page named ISO639-2, shouldn't it point to some page with just ISO639-1 codes to avoid confusion?
2) It should be part of some sentence
3) Please specify impact of this propery to default translation 

1) Corrected.
2) Yes, corrected.
3) Done.

File: src/xdocs/config_misc.xml [hunk 9]
```diff
@@ -328,6 +328,20 @@ messages.properties: Key 'ok' missing.
             <td><a href="property_types.html#string">String</a></td>
             <td><code>_</code></td>
           </tr>
+          <tr>
+             <td>requiredTranslations</td>
+             <td>
+                 Allows to specify language codes of required translations which must exist in project.
+                 Language code is composed of the lowercase, two-letter codes as defined by
+                 <a href="http://www.fatbellyman.com/webstuff/language_codes_639-1/">ISO 639-1</a>
```
Comment (mkordas) [inline src/xdocs/config_misc.xml]: very minor: missing dot at the end of sentence

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java [hunk 10]
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
...
+        };
+        verify(
+            createChecker(checkConfig),
+            propertyFiles,
+            getPath("checks"),
+            expected);
+    }
+
+    @Test
+    public void testTranslationFileAreMissing() throws Exception {
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java]: minor: testTranslationFile**s**AreMissing

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java [hunk 11]
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
...
+            IOException.class,
+            File.class);
         logIOException.setAccessible(true);
         logIOException.invoke(check, new IOException("test exception"), new File(""));
     }
 
+    @Test
+    public void testDefaultTranslationFileIsMissing() throws Exception {
+        final DefaultConfiguration checkConfig = createCheckConfig(TranslationCheck.class);
+        checkConfig.addAttribute("requiredTranslations", "ja, de, ja");
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java]: this test could include more complicated string set like ",ja,de,, ja," just to have full functionality coverage

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java [hunk 12]
```diff
@@ -120,10 +120,53 @@ public void testLogIOException() throws Exception {
         check.setMessageDispatcher(createChecker(checkConfig));
 
         Method logIOException = check.getClass().getDeclaredMethod("logIOException",
-                IOException.class,
-                File.class);
+            IOException.class,
```
Comment (mkordas) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheckTest.java]: looks like unnecessary changes in scope of this PR

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java [hunk 13]
```diff
@@ -112,12 +147,130 @@ protected void processFiltered(File file, List<String> lines) {
...
+            for (File currentFile : resourceBundle) {
+                final String currentFileName = currentFile.getName();
+                translationFileIsMissed =
+                    !currentFileName.contains('_' + languageCode + ".");
+                if (!translationFileIsMissed) {
+                    break;
+                }
+            }
+
+            if (translationFileIsMissed) {
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java]: `isMissing` is more common I think

**Comment by mkordas** [comment]
@MEZk, thanks a lot! I've added a few final, very minor comments to PR. 
@romani, I think you can start your code review and merge if no addtional issues will be found

**Comment by MEZk** [comment]
@mkordas 
>very minor: missing dot at the end of sentence

Corrected.

>minor: testTranslationFilesAreMissing

Corrected.

>this test could include more complicated string set like ,ja,de,, ja, just to have full functionality coverage

Added more complicated string set like ,ja,de,, ja.

>looks like unnecessary changes in scope of this PR

Corrected.

>isMissing is more common I think

Renamed.

P.S. Thanks. I appreciate your reviews.

**Comment by romani** [comment]
in second commit I still see "#" , please search for - "http://checkstyle.sourceforge.net/config.html\#Packages" , other comments will be later.

**Comment by MEZk** [comment]
@romani 

```http://checkstyle.sourceforge.net/config.html#Packages``` will be produced to user. ]
See [Test](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java#L189)

**Comment by romani** [comment]
ok, I miss escape symbol.

**Comment by MEZk** [comment]
@mkordas , @romani 

Added new option **requiredTranslations***
Added new option **translationsNotRequiredForBundleNamePattern**

**Comment by romani** [comment]
> translationsNotRequiredForBundleNamePattern

No, please go from positive logic and search only for "messages_*.properties" as format and do not touch any other property files.

**Comment by MEZk** [comment]
@romani 
Please explain why we should check only   "messages_*.properties". Oracle doesn't prohibit use of other names for translations bundles. See https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html

Or it is common practise to call translations bundles as "messages_*.properties" ?

**Comment by romani** [comment]
Let me investigate this, but main point is that *.properties files is general format to keep some string values , there is no limitations to use it for messages only. Right now more and more project try to use xml, json, yml for this still there are should be number of example where properties are still in use.

**Comment by romani** [comment]
@MEZk 

can you move translationsNotRequiredForBundleNamePattern to separate PR  ? to let other changes be applied.

**Comment by MEZk** [comment]
@romani 
Without translationsNotRequiredForBundleNamePattern we should provide translations for checkstyle-ant-task_es.properties

**Comment by MEZk** [comment]
@romani 
For now, I removed code connected with ```translationsNotRequiredForBundleNamePattern```
I'll create PR after you merge changes to avoid merge conflicts.

**Comment by romani** [comment]
Create an issue for this and keep it in you branch on remote

**Comment by mkordas** [comment]
@MEZk, how does it work right now? I see green builds without updating Checkstyle Ant Task properties.

**Comment by MEZk** [comment]
@mkordas 
See https://github.com/MEZk/checkstyle/commit/8a6af5a2cfa101e186bb158ab4e013858eab0ee0#diff-bfaac004d98e5f0f13a9b478ac5797cfR77

**Comment by mkordas** [comment]
Thanks, now I see. I think this PR is ready to merge. What dou you think @romani?

**Comment by romani** [comment]
some of items are not related changes , but lets consider them too, as we do serious update to Check

1)
> final Map<String, SortedSet<File>> propFilesMap =

Why not a 
http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Multimap.html
?
Why do we need SortedTree ? only to get default as first element ? I might miss smth , but sorted collection is too much for this logic.

2)
> if (!requiredTranslations.isEmpty()) {

no need to do this. remove that check and code become clear, and you will not do performance changes as "checkExistenceOfTranslationFiles" will not iterate and skip processing.

3)
> checkExistenceOfTranslationFilesInResourceBundle

name is too big. Please remove "OfTranslationFile" , as whole Check is about translations files

4)
> isMissing = !currentFileName.contains('_' + languageCode + ".");
> return String.format("%s_%s.properties",

"_" is hardcoded while we have option for it ?
please create UTs that would fail on this code and basenameSeparator is equal "-",  please attach links to that UTs when you put a "Done" on this point.
please  more ".properties" to variable.

5)
> private static String extractPropertyIdentifier(final File file,
            final String basenameSeparator) {

please remove final modifiers, do not pollute a code. I know that this decision is inconsistent with demanding "final" on local variables .... one day I will rethink its necessity to code, not now.

**Comment by MEZk** [comment]
@romani 
4) It is the ending format of property file name: [base name] + "_" + [language code] + [dot] + [extension].
For example:
```
messages-test + _ + ja + . + properties
```

or
````
messages + _ + ja + . + properties
````

See para 2: https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html
>the base name is followed by the language code

But yes, ```.properties``` should be replaced with **extension** option from base class.

**Comment by MEZk** [comment]
@romani 
1) 
>Why do we need SortedTree ? only to get default as first element ? I might miss smth , but sorted collection is too much for this logic.

Otherwise we should iterate though each element of the Set and check whether it is a default translation. Maybe we can use SortedSet just inside ```checkExistenceOfTranslationFilesInResourceBundle``` or try to change logic?

**Comment by romani** [comment]
> Maybe we can use SortedSet just inside checkExistenceOfTranslationFilesInResourceBundle or try to change logic?

It is make sense to use sorted collection only when you repetitive search over it few times. But our logic do search only ones for default - that is simple search that is required only ones. 

> So, It seams to me that basenameSeparator should be removed from check's options!

Just take a look at history (blame) and you will see a reason why it was added.

6)
> 
```
            boolean isMissing = false;
            for (File currentFile : resourceBundle) {
                final String currentFileName = currentFile.getName();
                isMissing = !currentFileName.contains('_' + languageCode + ".");
                if (!isMissing) {
                    break;
                }
            }
```

make is as function and name it let code read clearly.

**Comment by MEZk** [comment]
@romani 
Now I see why it is needed.
Are you ok with the following:

1) If we use basenameSeparator, a base name of resource bundle is right before separator.
For example:
For ```app-stage.properties``` base name is ```app```.
For ```app_stage.properties``` base name is ```app```.
For ```app-stage_de.properties``` base name is ```app```.
Because right now the base name is treated as it it described in the above examples.

2) The check looks only for messages bundles which names contain the word 'messages'.

**Comment by romani** [comment]
> 1) If we use basenameSeparator, a base name of resource bundle is right before separator.

all cases are right.

> 2) The check looks only for messages bundles which names contain the word 'messages' when it validates an existence of translations.

not sure why we need this decision in this PR, you already moved that to separate issue. Please explain.

**Comment by MEZk** [comment]
@romani 
1) Map has been replaced with SetMultimap from Guava. SortedSet has been replaced with Set.

2)
 >no need to do this. remove that check and code become clear, and you will not do performance changes as "checkExistenceOfTranslationFiles" will not iterate and skip processing.

Done. 

3) I think ```checkExistenceOfDefaultTranslation(filesInBundle)``` and  ```checkExistenceOfRequiredTranslations(filesInBundle)``` are better names for the checks for now. Do you agree? See [checkExistenceOfTranslations](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-bfaac004d98e5f0f13a9b478ac5797cfR163)

4) UTs: [testBaseNameSeparatorDefaultTranslationIsMissing](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-42885903150bfedec158bb760f2fe753R174) and [testBaseNameSeparatorTranslationsAreMissing](https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-42885903150bfedec158bb760f2fe753R195)

5) Done

>not sure why we need this decision in this PR, you already moved that to separate issue. Please explain.

Just to make sure that the check works fine without new option.

**In addition:**

null check was removed from  https://github.com/MEZk/checkstyle/commit/fa332a8decb58681505e46525b6a6d5344cf2786#diff-bfaac004d98e5f0f13a9b478ac5797cfL174 since SetMultimap.get(key) always returns a non-null, possibly empty collection. See [guava doc](https://github.com/google/guava/wiki/NewCollectionTypesExplained#multimap)

**Comment by romani** [comment]
please recheck this on JDK that Travis use - https://travis-ci.org/checkstyle/checkstyle/jobs/81554715#L103, latest jdk8 is http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html

"java version "1.8.0_31" vs "8u60"

mvn version is "3.2.5" https://travis-ci.org/checkstyle/checkstyle/jobs/81554715#L59

**Comment by MEZk** [comment]
@romani 
With maven 3.2.5.

|     | java 1.8.0.60| jav 1.8.0.31|
|---:|-----:|-----:|
|**mvn clean verify**| build success| build success|
|**mvn clean cobertura:clean cobertura:check**|build success | build success|
Strange.
 And AppVeyor uses java 1.8.0.45 and maven 3.2.5.

**Comment by romani** [comment]
at this issue nuance was in missed "clean"

**Comment by MEZk** [comment]
@romani 
The same problem

**Comment by romani** [comment]
Works fine on my laptop, and other linux boxes
```
$ mvn -version
Apache Maven 3.2.2 (45f7c06d68e745d05611f7fd14efb6594181933e; 2014-06-17T06:51:42-07:00)
Maven home: /opt/maven/default
Java version: 1.8.0_45, vendor: Oracle Corporation
Java home: /opt/jdk1.8.0_45/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.13.0-63-generic", arch: "amd64", family: "unix"
```

I presume that you run out of ideas and need help , right ?

**Comment by MEZk** [comment]
@romani 
Yes, I cannot imagine why Travis CI fails the build.

**Comment by romani** [comment]
I need your update to be part of 6.11:
- please create separate issue to resolve this coverage problem
- add additional commit to this PR and add exclude for coverage for this Check.

**Comment by romani** [comment]
Why you exclude it completely ? Put a certain covrrage on which travis is agree.

**Comment by MEZk** [comment]
@romani 
>please create separate issue to resolve this coverage problem

Done. See https://github.com/checkstyle/checkstyle/issues/2243 .

>add additional commit to this PR and add exclude for coverage for this Check.
Why you exclude it completely ? Put a certain covrrage on which travis is agree.

Done.

**Comment by romani** [comment]
merged as FF
